### PR TITLE
feat: capstone demo for v0.7.0 redirect-hook epic

### LINKED ISSUE #242: Capstone: redirect epic narrated demo and v0.7.0 release
## Goal

Produce a narrated `.mp4` that proves the redirect epic works end-to-end. Demonstrates before-state (raw `git diff` noise), the install-hooks ceremony, after-state redirect behavior, and the `review_change` one-call alternative. This is the non-negotiable capstone artifact for the epic.

Parent epic: #234. Depends on W2 / W3 / W4 / W5 / W6 (all implementation and docs merged).

## Demo Structure

Three segments, narrated, terminal-recorded via `creating-narrated-demos` skill:

### Segment 1: Before (the problem)
- Open a fresh shell with no hooks installed
- Run `git diff main..HEAD` against a representative branch in this repo (or a fixture)
- Show the noisy unified-diff output scrolling past
- Narration: "An agent reading this is going to spend half its context window on diff noise. There's a better tool, but the agent doesn't know."

### Segment 2: The fix
- Run `git-prism install-hooks --scope user`
- Show the resulting `~/.claude/settings.json` diff (just the new entries)
- Re-run `git diff main..HEAD`
- Hook fires: agent gets exit code 2 and the redirect message pointing at `get_change_manifest`
- Narration: "One command bundles the redirect logic that catches every variant — compound commands, subshells, variable substitution, the lot."

### Segment 3: One-call parity
- Demonstrate `review_change` from an MCP-aware agent (Claude Code session, scripted)
- Show the combined manifest + function_context payload
- Compare token count to the equivalent `git diff | git log -S` pipeline
- Narration: "review_change is the head-to-head competitor with `git diff`. Same intent, structured output, ~30% the tokens."

## Recording Pipeline

- Use existing `demo/demo.sh` + `demo/narration.md` pattern (extended for this epic — likely a new `demo/redirect-demo.sh` and `demo/redirect-narration.md`)
- ElevenLabs narration via the `creating-narrated-demos` skill
- Output: `demo/recordings/v0.7.0/redirect-epic.mp4`
- Also commit the `.cast` file (asciinema) for archival, even though the deliverable is `.mp4`

## Definition of Done

- [ ] Pre-flight: `env | grep -i ELEVEN` returns both `ELEVENLABS_API_KEY` and `ELEVENLABS_VOICE_ID`. If absent, block and ask the maintainer — do not substitute screenshots.
- [ ] `demo/redirect-demo.sh` script committed and runnable end-to-end against the current binary
- [ ] `demo/redirect-narration.md` written and aligned to demo timing
- [ ] `demo/recordings/v0.7.0/redirect-epic.mp4` rendered and committed (or stored per the existing convention if the repo doesn't track binaries)
- [ ] All three segments are visible in the final video
- [ ] At least one error/edge case shown (e.g., `git diff` in a subshell still firing the redirect)
- [ ] Demo is reproducible — running `demo/redirect-demo.sh` on a fresh machine produces the same flow
- [ ] Demo authenticity verified per `references/demo-authenticity.md`

## Out of Scope

- Browser-based UI demo (this is CLI-only)
- Localization (English only)
- A separate version-tagged release (the release happens after this issue closes — it is the gate)

## Notes

The capstone is the gate, not just the wrapper. If the demo can't be made to work convincingly, that's a signal that the implementation has a gap — fix the implementation before re-recording. Don't paper over with editing.

### PR DESCRIPTION
**Pull request by mikelane**
## Summary
- Capstone demo (`.mp4`) for the redirect-hook epic (#242), proving end-to-end: Claude's git muscle memory → hook interception → structured payoff
- `scripts/build-demo.py` — ElevenLabs TTS narration pipeline (stdlib-only Python)
- `scripts/test_build_demo.py` — 28 pytest tests covering all utility functions

## Demo narrative
Three segments tell one coherent story:
1. **The training problem** — Claude defaults to `git diff` because training on millions of git commands overrides MCP registration
2. **How the hook works** — `gh pr diff` hard-blocked (exit 2), `git diff` advisory (exit 0 + suggestion), benign commands silent (exit 0)
3. **The structured payoff** — `git-prism manifest` + `git-prism context` replace raw diffs with per-file metadata and blast-radius scoring

## Test plan
- [x] `python3 -m pytest scripts/test_build_demo.py -q` — 28 passed
- [x] `python3 scripts/build-demo.py demo/redirect-narration.md redirect-epic --output-dir demo/recordings/v0.7.0 --dry-run` — clean lint
- [x] `cargo test --bin git-prism` — 733 passed
- [x] `cargo clippy --all-targets -- -D warnings` — clean
- [x] `cargo fmt --check` — clean
- [x] `bash demo/redirect-demo.sh` — runs end-to-end with calibrated timing
- [x] `ffprobe demo/recordings/v0.7.0/redirect-epic.mp4` — both audio + video streams, 123.3s

Closes #242

🤖 Generated with [Claude Code](https://claude.ai/claude-code)

### CODE REVIEW

File: scripts/test_build_demo.py [hunk 1]
```diff
@@ -0,0 +1,358 @@
+"""Tests for build-demo.py utility functions."""
+import subprocess
```
Comment (mikelane) [inline scripts/test_build_demo.py]: **HIGH: Unused import `subprocess`** -- ruff F401. This import is never used in the test file. Remove it.

File: scripts/test_build_demo.py [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+from pathlib import Path
+
+import pytest
+
+# Add scripts/ to path so we can import build_demo
+import importlib.util
+import sys
```
Comment (mikelane) [inline scripts/test_build_demo.py]: **HIGH: Unused import `sys`** -- ruff F401. Remove it.

File: scripts/test_build_demo.py [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+
+_spec = importlib.util.spec_from_file_location(
+    "build_demo", Path(__file__).parent / "build-demo.py"
+)
+assert _spec is not None, "Could not locate build-demo.py next to test file"
+_mod = importlib.util.module_from_spec(_spec)  # type: ignore[arg-type]
```
Comment (mikelane) [inline scripts/test_build_demo.py]: **HIGH: Stale `type: ignore`** -- mypy reports `[unused-ignore]`. Remove `# type: ignore[arg-type]` from this line.

File: scripts/test_build_demo.py [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+
+    assert len(result) == 1
+    assert result[0]["name"] == "real"
+
+
+# ---------------------------------------------------------------------------
+# lint_segments
+# ---------------------------------------------------------------------------
+
+def test_lint_segments_returns_zero_for_clean_text(capsys: pytest.CaptureFixture) -> None:
```
Comment (mikelane) [inline scripts/test_build_demo.py]: **MEDIUM: Missing type parameter** -- mypy reports `[type-arg]` for `CaptureFixture`. Use `CaptureFixture[str]` (all 6 instances of this in the test file).

File: scripts/test_build_demo.py [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+    assert not output.exists(), (
+        "output file must not exist when ffmpeg fails"
+    )
+
+
+# ---------------------------------------------------------------------------
+# generate_playwright_skeleton
+# ---------------------------------------------------------------------------
+
+def _make_timing(names_and_durations: list[tuple[str, float]]) -> list[dict]:
```
Comment (mikelane) [inline scripts/test_build_demo.py]: **MEDIUM: Missing type parameter** -- mypy reports `[type-arg]` for `dict`. Use `list[dict[str, str | float]]` instead of bare `list[dict]`.

File: demo/redirect-narration.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
+<!-- Narration script for git-prism v0.7.0 redirect-hook capstone demo (issue #242).
+     Processed by scripts/build-demo.py — uses <!-- SEGMENT: name --> markers.
+     TTS-friendly: no markdown syntax, no shell special characters in segment bodies.
+     Segment names map 1-to-1 to sleep-calibrated sections in demo/redirect-demo.sh. -->
+
+<!-- SEGMENT: problem_intro -->
+I've got a notes file in a git repository here — tracked, committed, under version control. Now watch what happens when an agent runs an unguarded bash redirect.
```
Comment (mikelane) [inline demo/redirect-narration.md]: **MEDIUM: Narration text mismatch with timing.json.**

The text in this segment differs from what was actually sent to ElevenLabs (recorded in `redirect-epic_timing.json`). This file says:

> "I've got a notes file in a git repository here -- tracked, committed, under version control. Now watch what happens when an agent runs an unguarded bash redirect."

But timing.json records the TTS text as:

> "Here is a git repository with a notes file containing research I want to keep. Under version control, tracked, committed. Now watch what happens when an agent runs an unguarded bash redirect."

This affects ALL 10 segments -- only `hook_install` matches exactly. If someone reruns `build-demo.py` with this narration.md, the generated audio will produce different durations than the sleep values in `redirect-demo.sh`.

Either update narration.md to match the timing.json text (what was actually TTS'd), or add a comment noting the discrepancy. See PR description for the full cross-check analysis.

File: demo/recordings/v0.7.0/.gitkeep [hunk 7]
```diff
@@ -0,0 +1,8 @@
+# Recordings for the v0.7.0 redirect-hook capstone demo land here.
+#
+# Expected artifacts after `asciinema rec` + render pipeline:
+#   - git-prism-redirect-demo.cast
```
Comment (mikelane) [inline demo/recordings/v0.7.0/.gitkeep]: **MEDIUM: Filename references are wrong.**

The comment here references `git-prism-redirect-demo.{cast,gif,mp4}` but the actual artifacts are named `redirect-epic.{cast,gif,mp4}`. Update the comment to match the real filenames, or rename the files to match the comment.

File: scripts/build-demo.py [hunk 8]
```diff
@@ -0,0 +1,498 @@
...
+        headers={
+            "Accept": "audio/mpeg",
+            "Content-Type": "application/json",
+            "xi-api-key": api_key,
+        },
+        method="POST",
+    )
+
+    try:
+        with urlopen(request) as response:  # noqa: S310
```
Comment (mikelane) [inline scripts/build-demo.py]: **LOW: `noqa: S310` would benefit from an explanatory comment.**

Bandit rule S310 flags `urlopen` for potential URL injection. Here the URL is a hardcoded ElevenLabs API endpoint, so it's safe. Adding a comment like `# Safe: URL is a hardcoded ElevenLabs API endpoint, no user input` would help future readers understand why the suppression exists.

**Review by mikelane — COMMENTED** [review]
## Review: PR #253 -- capstone demo for v0.7.0 redirect-hook epic

### Overall assessment: Changes needed before merge

No CRITICAL blockers, but several HIGH and MEDIUM findings that should be addressed. All are mechanical fixes -- no redesign needed.

---

### Summary of findings

**HIGH (3 findings)**
1. **Ruff F401: unused import `subprocess`** in `scripts/test_build_demo.py:2` -- trivially fixable (delete the line)
2. **Ruff F401: unused import `sys`** in `scripts/test_build_demo.py:9` -- trivially fixable (delete the line)
3. **Mypy `[unused-ignore]`** on `test_build_demo.py:15` -- stale `# type: ignore[arg-type]` suppress comment. Remove it.

**MEDIUM (3 findings)**
4. **Narration text mismatch across ALL 10 segments** -- `demo/redirect-narration.md` and `demo/recordings/v0.7.0/redirect-epic_timing.json` disagree on what text was actually sent to ElevenLabs. Only `hook_install` matches exactly. This breaks the reproducibility contract: running `build-demo.py` with the current narration.md would produce audio with different durations than the sleep values in `redirect-demo.sh`. Fix by syncing narration.md to timing.json (or vice versa).
5. **Wrong artifact filenames in `.gitkeep`** -- `demo/recordings/v0.7.0/.gitkeep` references `git-prism-redirect-demo.{cast,gif,mp4}` but actual files are `redirect-epic.*`.
6. **Mypy `[type-arg]` errors** (7 instances) -- missing generic type parameters on `CaptureFixture` (6x) and bare `dict` return type (1x). Use `CaptureFixture[str]` and `list[dict[str, str | float]]`.

**LOW (1 finding)**
7. `noqa: S310` on `build-demo.py:156` would benefit from a comment explaining why the suppression is safe.

---

### What's good

- **build-demo.py architecture is clean.** Clear pipeline stages (parse -> generate -> measure -> concatenate), well-named functions, good error messages. The `main()` decomposition into `generate_audio_clips()` and `write_timing_and_summary()` subroutines is exactly the right abstraction level.
- **Test coverage for the build tool is strong.** 28 tests covering segment parsing (empty/no markers/single/multiple/boundaries/comment-stripping), linting (clean/comment/heading/inline/counting), concatenation (cleanup on failure), and Playwright skeleton generation (file creation/buffer/timeout/name/url embedding).
- **Bash script safety.** `set -euo pipefail`, proper `trap` cleanup, preflight checks for binary and hook script existence, dynamic repo-root resolution via `SCRIPT_DIR`/`REPO_ROOT`.
- **Shellcheck passes clean** (exit 0).
- **Mypy on build-demo.py passes clean** under `--strict`.
- **`validate_demo_name` argparse validator** is a good defense against injection.
- **Timing-calibrated demo script** with well-documented sleep values -- each sleep maps to a specific narration segment with the narration duration noted.

---

### Items verified clean

- No `|| true` in CI workflows (the 4 instances in `redirect-demo.sh` are in demo context with clear explanatory comments)
- No production code changes -- zero blast radius beyond `demo/` and `scripts/`
- `gh pr view --json mergeStateStatus` reports `CLEAN` (no merge conflicts)
- PR is `MERGEABLE`
- All 28 pytest tests pass

---

### Recommendation

The 3 HIGH findings (unused imports + stale type:ignore) are one-commit fixes. The narration mismatch (finding 4) is the most substantively important -- it's a latent correctness issue that affects anyone trying to reproduce or iterate on this demo. Fix the HIGH items, address the narration mismatch (even if just with a comment acknowledging the discrepancy), and the remaining MEDIUM/LOW items, then merge.

File: scripts/test_build_demo.py [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+import pytest
+
+import importlib.util
+
+_spec = importlib.util.spec_from_file_location(
+    "build_demo", Path(__file__).parent / "build-demo.py"
+)
+assert _spec is not None, "Could not locate build-demo.py next to test file"
+_mod = importlib.util.module_from_spec(_spec)
+_spec.loader.exec_module(_mod)
```
Comment (mikelane) [inline scripts/test_build_demo.py:14]: **CRITICAL**: `_spec.loader` may be `None`. The code asserts `_spec is not None` on line 11, but does not assert that `_spec.loader` is not None before calling `.exec_module()`. Add `assert _spec.loader is not None, "Cannot locate build-demo.py loader"` above line 13.

File: scripts/test_build_demo.py [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+    warning_count = lint_segments(segments)
+    stderr = capsys.readouterr().err
+
+    assert warning_count == 1
+    assert "bad" in stderr
+    assert "heading" in stderr
+
+
+def test_lint_segments_info_on_markdown_inline_syntax_does_not_increment_warning_count(
+    capsys: pytest.CaptureFixture,
```
Comment (mikelane) [inline scripts/test_build_demo.py]: **HIGH**: Missing type parameter for generic type `CaptureFixture`. Use `pytest.CaptureFixture[str]` to match the annotation used in the other lint tests (lines 176, 184, 195).

File: scripts/test_build_demo.py [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+
+    warning_count = lint_segments(segments)
+    stderr = capsys.readouterr().err
+
+    assert warning_count == 0
+    assert "INFO" in stderr
+
+
+def test_lint_segments_counts_multiple_warnings_across_segments(
+    capsys: pytest.CaptureFixture,
```
Comment (mikelane) [inline scripts/test_build_demo.py]: **HIGH**: Missing type parameter for generic type `CaptureFixture`. Use `pytest.CaptureFixture[str]`.

File: scripts/test_build_demo.py [hunk 12]
```diff
@@ -0,0 +1,355 @@
...
+        {"name": "second", "text": "# Heading in body"},
+    ]
+
+    warning_count = lint_segments(segments)
+
+    assert warning_count == 2
+
+
+def test_lint_segments_counts_multiple_warnings_within_one_segment(
+    capsys: pytest.CaptureFixture,
```
Comment (mikelane) [inline scripts/test_build_demo.py]: **HIGH**: Missing type parameter for generic type `CaptureFixture`. Use `pytest.CaptureFixture[str]`.

File: scripts/test_build_demo.py (hunk 10 above)
Comment (mikelane) [inline scripts/test_build_demo.py]: **HIGH**: Bare `pytest.CaptureFixture` type annotation missing the type parameter. Other tests (lines 176, 184, 195) correctly use `pytest.CaptureFixture[str]`. This function also uses `pytest.CaptureFixture[str]` (it reads stderr via `capsys.readouterr()`). Change to `capsys: pytest.CaptureFixture[str]`.

File: scripts/test_build_demo.py (hunk 11 above)
Comment (mikelane) [inline scripts/test_build_demo.py]: **HIGH**: Bare `pytest.CaptureFixture` type annotation missing the type parameter. Change to `capsys: pytest.CaptureFixture[str]` to match the pattern used elsewhere in this file.

File: scripts/test_build_demo.py (hunk 12 above)
Comment (mikelane) [inline scripts/test_build_demo.py]: **HIGH**: Bare `pytest.CaptureFixture` type annotation missing the type parameter. Change to `capsys: pytest.CaptureFixture[str]` to match the pattern used elsewhere in this file.

**Review by mikelane — COMMENTED** [review]
## Re-review of PR #253 (commit 1ab5bdc)

Thank you for the fix commit. Five of the five original findings are confirmed resolved:

- **Confirmed fixed**: Unused `subprocess`/`sys` imports removed
- **Confirmed fixed**: All `type: ignore` comments removed (zero matches found)
- **Confirmed fixed**: `noqa: S310` replaced with explicit `timeout=30` on `urlopen` call
- **Confirmed fixed**: Narration text in `redirect-narration.md` matches `redirect-epic_timing.json` exactly across all 10 segments
- **Partially fixed**: Type annotations — `list[dict[str, float | str]]` applied in production code, and 3 of 6 `CaptureFixture[str]` annotations applied in tests

### Blocker: mypy fails with 4 errors

```
scripts/test_build_demo.py:13: error: Item "None" of "Loader | None" has no attribute "exec_module"  [union-attr]
scripts/test_build_demo.py:207: error: Missing type parameters for generic type "CaptureFixture"  [type-arg]
scripts/test_build_demo.py:221: error: Missing type parameters for generic type "CaptureFixture"  [type-arg]
scripts/test_build_demo.py:234: error: Missing type parameters for generic type "CaptureFixture"  [type-arg]
```

Three tests (lines 207, 221, 234) still use bare `pytest.CaptureFixture` without the `[str]` type parameter, despite this being flagged in the previous review and partially addressed on lines 176, 184, and 195. The fourth error is a null-safety gap: `_spec.loader` is typed as `Loader | None` and no assertion guards the `.exec_module()` call.

Each fix is one line. Add `assert _spec.loader is not None` above line 13, and add `[str]` to the three `CaptureFixture` annotations on lines 207, 221, 234.

### What passes clean

- ruff: zero findings
- pytest: 28/28 passing (0.15s)
- clippy + fmt + test in pre-push hook: all passed (confirmed in commit message)

### Verdict: Needs changes before merge

All five fixes from the previous review are either fully or partially addressed. The remaining issues are the three bare `CaptureFixture` annotations that were only partially corrected and one missing null-guard. All four are mechanical one-line changes.

**Review by mikelane — COMMENTED** [review]
## Re-review Verdict: REQUEST CHANGES

Thanks for the quick turnaround on the previous findings. Five of six issues from the first review are indeed resolved. However, the mypy strict pass was incomplete.

## Confirmed Fixes

1. **Unused imports** `subprocess` and `sys` in `test_build_demo.py` -- removed.
2. **Stale `type: ignore` comments** -- removed.
3. **Narration text mismatch** -- `redirect-narration.md` now matches `redirect-epic_timing.json` text verbatim, which is the canonical source since audio was recorded from it. Verified all 10 segment texts character by character.
4. **`# noqa: S310`** -- replaced with explicit `timeout=30` on `urlopen` call.
5. **Bare type annotations** -- partially fixed (3 of 6 `CaptureFixture` now carry `[str]`).

## Blockers

### Mypy strict fails with 4 errors in `scripts/test_build_demo.py`

**Line 13**: `_spec.loader` may be `None`. The code asserts `_spec is not None` but never asserts `_spec.loader is not None` before calling `.exec_module()`. In practice loader would never be None for this path, but mypy strict mode requires the guard. Add `assert _spec.loader is not None` above line 13.

**Lines 207, 221, 234**: Bare `pytest.CaptureFixture` type annotations. The first three `lint_segments` tests (lines 176, 184, 195) correctly use `pytest.CaptureFixture[str]`, but the remaining three reverted to the bare generic. Add `[str]` to all three.

## What Looks Good

- `build-demo.py` is cleanly structured with excellent docstrings and type annotations. The function decomposition into parse/lint/generate/concatenate/write phases is well done.
- `validate_demo_name` provides real defense against path-injection attacks on file names.
- `concatenate_audio` properly cleans up `concat_list.txt` on all exit paths (success, ffmpeg failure, ffmpeg not found).
- `redirect-demo.sh` sleep values are calibrated from the timing JSON and the sandbox trap cleanup is well-handled.
- `parse_segments` regex boundary detection is carefully documented with the exact boundary conditions enumerated.
- All 28 tests pass, and the test design is strong -- tests exercise public API only, use parametrization appropriately, and include informative assertion messages.
- Ruff passes cleanly with no issues.

## Decision

REQUEST CHANGES. The two mypy issues (4 errors) must be resolved before this can merge. These are the same class of finding flagged in the first review and should be easy to fix -- one `assert` statement and three `[str]` type parameters.

**Review by mikelane — COMMENTED** [review]
## Re-review Verdict: REQUEST CHANGES

Thanks for the quick turnaround on the previous findings. Five of six issues from the first review are indeed resolved. However, the mypy strict pass was incomplete.

## Confirmed Fixes

1. **Unused imports** `subprocess` and `sys` in `test_build_demo.py` -- removed.
2. **Stale `type: ignore` comments** -- removed.
3. **Narration text mismatch** -- `redirect-narration.md` now matches `redirect-epic_timing.json` text verbatim, which is the canonical source since audio was recorded from it. Verified all 10 segment texts character by character.
4. **`# noqa: S310`** -- replaced with explicit `timeout=30` on `urlopen` call.
5. **Bare type annotations** -- partially fixed (3 of 6 `CaptureFixture` now carry `[str]`).

## Blockers

### Mypy strict fails with 4 errors in `scripts/test_build_demo.py`

**Line 13**: `_spec.loader` may be `None`. The code asserts `_spec is not None` but never asserts `_spec.loader is not None` before calling `.exec_module()`. In practice loader would never be None for this path, but mypy strict mode requires the guard. Add `assert _spec.loader is not None` above line 13.

**Lines 207, 221, 234**: Bare `pytest.CaptureFixture` type annotations. The first three `lint_segments` tests (lines 176, 184, 195) correctly use `pytest.CaptureFixture[str]`, but the remaining three reverted to the bare generic. Add `[str]` to all three.

## What Looks Good

- `build-demo.py` is cleanly structured with excellent docstrings and type annotations. The function decomposition into parse/lint/generate/concatenate/write phases is well done.
- `validate_demo_name` provides real defense against path-injection attacks on file names.
- `concatenate_audio` properly cleans up `concat_list.txt` on all exit paths (success, ffmpeg failure, ffmpeg not found).
- `redirect-demo.sh` sleep values are calibrated from the timing JSON and the sandbox trap cleanup is well-handled.
- `parse_segments` regex boundary detection is carefully documented with the exact boundary conditions enumerated.
- All 28 tests pass, and the test design is strong -- tests exercise public API only, use parametrization appropriately, and include informative assertion messages.
- Ruff passes cleanly with no issues.

## Decision

REQUEST CHANGES. The two mypy issues (4 errors) must be resolved before this can merge. These are the same class of finding flagged in the first review and should be easy to fix -- one `assert` statement and three `[str]` type parameters.

File: demo/recordings/v0.7.0/.gitkeep (hunk 7 above)
Comment (mikelane) [inline demo/recordings/v0.7.0/.gitkeep]: The expected artifact filenames (`git-prism-redirect-demo.cast`, `.gif`, `.mp4`) do not match the actual files (`redirect-epic.cast`, `.gif`, `.mp4`). Either the filenames in the repo, or these documentation references, should be aligned.

**Review by mikelane — COMMENTED** [review]
## Summary

Clean capstone demo PR for the v0.7.0 redirect-hook epic (#242). The three-segment narrative ("git muscle memory" → "hook mechanics" → "structured payoff") is coherent and well-constructed. The ElevenLabs TTS pipeline (`scripts/build-demo.py`) has thorough error handling and excellent test coverage (28 tests). Sleep values in `demo/redirect-demo.sh` are correctly calibrated against `redirect-epic_timing.json` within rounding tolerance.

## Findings

**LOW: Orphan clip files** — `redirect-epic_problem_intro.mp3` (~200KB) and `redirect-epic_problem_clobber.mp3` (~260KB) in `demo/recordings/v0.7.0/clips/` have no corresponding `<!-- SEGMENT: problem_intro -->` or `<!-- SEGMENT: problem_clobber -->` marker in the current `demo/redirect-narration.md`. They are not referenced by `redirect-epic_timing.json` or the demo script. These are stale artifacts from the pre-rewrite narration draft. Removing them keeps the repo clean and avoids confusion.

**LOW: Stale .gitkeep artifact names** — `demo/recordings/v0.7.0/.gitkeep` line 4 references expected filenames `git-prism-redirect-demo.cast/.gif/.mp4` but the actual files are named `redirect-epic.*`. The comment should match reality.

## What Passed

- Python: 28/28 tests (pytest), mypy --strict clean on both build-demo.py and test_build_demo.py
- Rust: 733/733 tests, clippy --all-targets clean, cargo fmt --check clean
- Bash: syntax check clean (`bash -n`), set -eo pipefail correctly handled with `|| true` on head-truncated pipelines
- Security: API key from env only, urlopen timeout=30, no hardcoded secrets, no eval/exec on user input
- No suppressions found: zero `noqa`, `type: ignore`, or linter bypass comments in any file
- Architecture: clean main() decomposition, single-responsibility functions, no speculative generality
- Narration text is TTS-friendly (no markdown syntax in segment bodies per the dry-run output)

## Praise

- The comment on every `|| true` in the bash script explaining WHY it is there (SIGPIPE from head truncation, expected failure of status check) is exactly the right pattern — it prevents future maintainers from "cleaning up" what looks like a workaround.
- `validate_demo_name` as an argparse type function is a clean pattern that keeps validation at the parsing boundary.
- The `_spec.loader is not None` guard (from previous review feedback) shows the fix landed correctly.

## Verdict

No CRITICAL or HIGH blockers. The two LOW findings are cosmetic cleanup (orphan binary files and stale comment references) and do not affect correctness or functionality. The PR is ready.

**Review by mikelane — COMMENTED** [review]
## Summary

Clean capstone demo PR for the v0.7.0 redirect-hook epic (#242). The three-segment narrative ("git muscle memory" → "hook mechanics" → "structured payoff") is coherent and well-constructed. The ElevenLabs TTS pipeline (`scripts/build-demo.py`) has thorough error handling and excellent test coverage (28 tests). Sleep values in `demo/redirect-demo.sh` are correctly calibrated against `redirect-epic_timing.json` within rounding tolerance.

## Findings

**LOW: Orphan clip files** — `redirect-epic_problem_intro.mp3` (~200KB) and `redirect-epic_problem_clobber.mp3` (~260KB) in `demo/recordings/v0.7.0/clips/` have no corresponding `<!-- SEGMENT: problem_intro -->` or `<!-- SEGMENT: problem_clobber -->` marker in the current `demo/redirect-narration.md`. They are not referenced by `redirect-epic_timing.json` or the demo script. These are stale artifacts from the pre-rewrite narration draft. Removing them keeps the repo clean and avoids confusion.

**LOW: Stale .gitkeep artifact names** — `demo/recordings/v0.7.0/.gitkeep` line 4 references expected filenames `git-prism-redirect-demo.cast/.gif/.mp4` but the actual files are named `redirect-epic.*`. The comment should match reality.

## What Passed

- Python: 28/28 tests (pytest), mypy --strict clean on both build-demo.py and test_build_demo.py
- Rust: 733/733 tests, clippy --all-targets clean, cargo fmt --check clean
- Bash: syntax check clean (`bash -n`), set -eo pipefail correctly handled with `|| true` on head-truncated pipelines
- Security: API key from env only, urlopen timeout=30, no hardcoded secrets, no eval/exec on user input
- No suppressions found: zero `noqa`, `type: ignore`, or linter bypass comments in any file
- Architecture: clean main() decomposition, single-responsibility functions, no speculative generality
- Narration text is TTS-friendly (no markdown syntax in segment bodies per the dry-run output)

## Praise

- The comment on every `|| true` in the bash script explaining WHY it is there (SIGPIPE from head truncation, expected failure of status check) is exactly the right pattern — it prevents future maintainers from "cleaning up" what looks like a workaround.
- `validate_demo_name` as an argparse type function is a clean pattern that keeps validation at the parsing boundary.
- The `_spec.loader is not None` guard (from previous review feedback) shows the fix landed correctly.

## Verdict

No CRITICAL or HIGH blockers. The two LOW findings are cosmetic cleanup (orphan binary files and stale comment references) and do not affect correctness or functionality. The PR is ready.

Decision: APPROVED
